pkg/kubelet/kubelet_pods.go

```diff
@@ -1138,28 +1135,44 @@ func (kl *Kubelet) HandlePodCleanups(ctx context.Context) error {
-		case ok && workerState == SyncPod, ok && workerState == TerminatingPod:
-			// if the pod worker is already in charge of this pod, we don't need to do anything
-			continue
-		default:
-			// If the pod isn't in the set that should be running and isn't already terminating, terminate
-			// now. This termination is aggressive because all known pods should already be in a known state
-			// (i.e. a removed static pod should already be terminating), so these are pods that were
-			// orphaned due to kubelet restart or bugs. Since housekeeping blocks other config changes, we
-			// know that another pod wasn't started in the background so we are safe to terminate the
-			// unknown pods.
-			if _, ok := allPodsByUID[runningPod.ID]; !ok {
-				klog.V(3).InfoS("Clean up orphaned pod containers", "podUID", runningPod.ID)
-				one := int64(1)
-				kl.podWorkers.UpdatePod(UpdatePodOptions{
-					UpdateType: kubetypes.SyncPodKill,
-					RunningPod: runningPod,
-					KillPodOptions: &KillPodOptions{
-						PodTerminationGracePeriodSecondsOverride: &one,
-					},
-				})
+		// If there are runtime pod elements in CRI that are unknown to the pod worker, terminate them
+		// now. Since housekeeping is synchronous to other pod worker updates, we know that no pods have
+		// been added to the pod worker in the meantime. Note that pods that are not visible in the runtime
+		// but which were previously known are terminated by SyncKnownPods above.
+		_, knownPod := workingPods[runningPod.ID]
+		if !knownPod {
+			one := int64(1)
+			killPodOptions := &KillPodOptions{
+				PodTerminationGracePeriodSecondsOverride: &one,
 			}
+			klog.V(3).InfoS("Clean up containers for orphaned pod we had not seen before", "podUID", runningPod.ID, "killPodOptions", killPodOptions)
+			kl.podWorkers.UpdatePod(UpdatePodOptions{
+				UpdateType:     kubetypes.SyncPodKill,
+				RunningPod:     runningPod,
+				KillPodOptions: killPodOptions,
+			})
+
+			// the running pod is now known as well
+			workingPods[runningPod.ID] = TerminatingPod
+		}
+	}
+
+	// At this point, the pod worker is now aware of which pods are not desired (SyncKnownPods) and
+	// which pods may be lingering (from the runtimeCache). We now look through the set of desired
+	// pods for those that the pod worker is not aware of and deliver an update. Each pod added here
+	// indicates a potential problem in synchronization in the kubelet and should be investigated, but
+	// that is the role of reconciliation in general.
+	for _, desiredPod := range allPods {
+		_, knownPod := workingPods[desiredPod.UID]
+		if !knownPod {
+			klog.V(3).InfoS("Notify the pod worker of a pod that was not detected, potential configuration race", "podUID", desiredPod.UID)
+			kl.podWorkers.UpdatePod(UpdatePodOptions{
+				UpdateType: kubetypes.SyncPodCreate,
+				Pod:        desiredPod,
+			})
+
+			// the desired pod is now known as well
+			workingPods[desiredPod.UID] = SyncPod
 		}
```

Comment:
Think about it this way:

1. Two components need to get in sync, one upstream of the other
2. Kubelet config creates "desiredPods" in pod manager
3. HandlePodAdditions is the incremental informer
4. If there's a bug, an addition is missed (unlikely, but we don't actually check anywhere else)
5. HandlePodCleanups is the resync - should compare `desired-actual` (syncKnownPods), `actual-desired` (here), and `runtime-actual` (`UpdatePod({kill RunningPod: ...}}` above)

The case here is exceptional, i don't expect to hit it, but we should.  This is a strong indicator of a bug elsewhere in kubelet, as unlikely as it might be, so this is "sign of programmer error".

Reply:
I see, that makes sense. 

One question though -- as I understand the pod is only added to the pod manager in `HandlePodAdditions` AFAIU (see https://github.com/kubernetes/kubernetes/blob/1749bb29916682843c120cd4f69a1a610b3b3038/pkg/kubelet/kubelet.go#L2341)

So if for whatever reason we missed the update in `HandlePodAdditions` then I assume the pod would not be tracked by the pod manager (i.e. the "desired set"). In that case, it seems like this logic wouldn't help us because it depends on the pod manager being being up to date in the first place? 

Or is the scenario we are trying to catch is that `HandlePodAdditions` was called, it was added to pod manager there, but for whatever reason (bug, etc...) pod worker did not become aware of it? I have no issue with keeping the logic, just want to understand it a bit deeper :)